propertyestimator/utils/quantities.py

```diff
@@ -0,0 +1,656 @@
+    another whose source is a trajectory file name.
+
+    Examples
+    --------
+    To add two **independent** quantities together:
+
+    >>> from simtk import unit
+    >>>
+    >>> a_value = 5 * unit.angstrom
+    >>> a_uncertainty = 0.03 * unit.angstrom
+    >>>
+    >>> b_value = 10 * unit.angstrom
+    >>> b_uncertainty = 0.04 * unit.angstrom
+    >>>
+    >>> quantity_a = TaggedEstimatedQuantity(a_value, a_uncertainty, 'calc_325262315:npt_production')
+    >>> quantity_b = TaggedEstimatedQuantity(b_value, b_uncertainty, 'calc_987234582:npt_production')
+    >>>
+    >>> quantity_addition = quantity_a + quantity_b
+
+    To subtract one **independent** quantity from another:
+
+    >>> quantity_subtraction = quantity_b - quantity_a
+
+    Attempting to add quantities from the same source (i.e probably correlated)
+    will raise a `DependantValuesException`.
+
+    >>> c_value = 5.04 * unit.angstrom
+    >>> c_uncertainty = 0.029 * unit.angstrom
+    >>>
+    >>> quantity_c = TaggedEstimatedQuantity(c_value, c_uncertainty, 'calc_325262315:npt_production')
+    >>>
+    >>> # The below raised a DependantValuesException.
+    >>> quantity_addition = quantity_a + quantity_c
+    """
+
+    @property
+    def source(self):
+        return self._source
+
+    def __init__(self, value, uncertainty, source):
+        """Constructs a new TaggedEstimatedQuantity object.
+
+        Parameters
+        ----------
+        value: unit.Quantity
+            The value of the estimated quantity.
+        uncertainty: unit.Quantity
+            The uncertainty in the value of the estimated quantity.
+        source: str
+            A string representation of where this value came from. This
+            value is employed whenever this object is involved in any
+            mathematical operations to ensure values from the same source
+            are not accidentally combined.
+
+            An example of this may be the file path of the trajectory from
+            which this value was derived, or the name of the workflow protocol
+            which calculated it.
+        """
+        super().__init__(value, uncertainty)
+
+        self._source = source
+
+    def __add__(self, other):
+
+        assert isinstance(other, TaggedEstimatedQuantity)
+
+        if self.source == other.source:
+            raise DependantValuesException()
+
+        return super(TaggedEstimatedQuantity, self).__add__(other)
+
+    def __sub__(self, other):
+
+        assert isinstance(other, TaggedEstimatedQuantity)
+
+        if self.source == other.source:
+            raise DependantValuesException()
+
+        return super(TaggedEstimatedQuantity, self).__sub__(other)
+
+
+class ExplicitEstimatedQuantity:
+    """An implementation of an `EstimatedQuantity` class, which is almost identical
+    to the VanillaEstimatedQuantity implementation, except now all operators have been
+    replaced with explicit method calls.
+
```

Comment:
@SimonBoothroyd : Could you add a line or two about when we should use this vs `VanillaEstimatedQuantity`?